### ISSUE TITLE
Dual-Directional Model

### ISSUE THREAD

**Issue by csmangum** [csmangum/GCA#1]
Can a model be trained to learn a system forward and backward? Essentially capable of going forward and backward in "time" at will?

**Comment by csmangum** [comment]
You can design a model that learns to predict in both directions, using separate pathways or mechanisms for forward and backward predictions. This could mean having one part of the network dedicated to forward prediction and another to backward prediction, each trained on their respective tasks. The rule encoding could be used as part of the input in both cases, or the model could be designed to infer it in the backward direction.

A dual-directional model is designed to handle tasks that require understanding or predicting information in two opposite directions, such as sequences where both past and future contexts are important, or problems where both forward and backward inference are needed. In the context of predicting the state of an object based on a rule and also inferring previous states and rules from a given state, a dual-directional model can be specifically tailored to learn these forward and backward dynamics concurrently.

### Structure of a Dual-Directional Model

The structure of a dual-directional model typically involves separate components or pathways for processing information in each direction:

- **Forward Pathway:** This part of the model is responsible for predicting the next state of an object based on its current state and an encoding of the rule. It's essentially a predictive model that maps a current state to a future state, considering the governing rule.

- **Backward Pathway:** The backward pathway is tasked with the inverse problem – given a state or a sequence of states, it aims to infer the previous state and potentially the rule that led to the current state. This requires the model to understand the dynamics of the system in reverse.

### Training Dual-Directional Models

Training such a model involves presenting it with data that enables learning both forward and backward relationships. This typically requires:

- **Dual Objectives:** The model is optimized for two objectives simultaneously. The forward pathway is trained to minimize the discrepancy between its predictions and the true next states. Meanwhile, the backward pathway is trained to accurately infer previous states and the rules applied, minimizing errors in backward inference.

- **Shared Representations:** While the pathways are distinct, they might share some underlying representations or layers. This allows the model to leverage common features between forward and backward predictions, making the model more efficient and potentially improving its performance by learning a more comprehensive representation of the state space.

### Applications and Advantages

Dual-directional models are particularly useful in scenarios where understanding the bidirectional context or dynamics is crucial. For example:

- **Sequence Prediction:** In natural language processing, understanding both past and future context can be crucial for tasks like translation, summarization, or question answering.
- **Time Series Analysis:** Predicting future values based on past data and vice versa, for example, in financial modeling or weather forecasting.
- **Physical and Biological Systems:** Understanding the dynamics of systems where the future state depends on previous states and where inferring past conditions from current observations is valuable.

### Advantages

- **Versatility:** Can handle complex tasks that require understanding or predicting in both directions.
- **Efficiency:** By potentially sharing representations, the model can learn more generalized features of the data, improving its predictive capabilities.
- **Improved Understanding:** Can provide insights not just into the future or past states but also into the underlying rules or dynamics governing the system.

### Implementation Considerations

- **Architecture Design:** The design of the forward and backward pathways must be tailored to the specific characteristics of the problem, including the nature of the states and the rules governing transitions.
- **Data Requirements:** Sufficient and appropriately structured data is needed to train both aspects of the model effectively.
- **Complexity and Resources:** Managing the complexity and computational resources required for training and deploying dual-directional models, especially for large or complex state spaces.

In summary, a dual-directional model offers a powerful approach for simultaneously addressing forward and backward prediction tasks, leveraging the strengths of neural networks to capture complex relationships within data. Its implementation, while potentially resource-intensive, opens up new possibilities for predictive modeling and inference in a wide range of applications.

**Comment by csmangum** [comment]
Implementing a dual-directional model for elementary cellular automata (CA) in PyTorch involves creating a neural network that can both predict the next state of a cell based on its current state and two neighbors (forward direction) and infer the previous state of a cell from a sequence of states, as well as identify the CA rule used (backward direction). Here's a high-level approach to designing and implementing such a model:

### Model Architecture

1. **Forward Pathway:**
   - **Input:** The current state of the cell and its two immediate neighbors. This can be represented as a 3-bit vector.
   - **Network:** A small feedforward neural network (or potentially a more complex architecture if you're dealing with a broader context or additional features) that takes this 3-bit input and predicts the next state of the cell (1 or 0).

2. **Backward Pathway:**
   - **Input:** A sequence of states for a cell. The length of this sequence depends on your specific requirements and the complexity of inferring the previous state and the CA rule.
   - **Encoder:** A component (e.g., LSTM, GRU, or Transformer encoder) that processes the sequence of states and encodes it into a fixed-size vector representation.
   - **Decoder:** Two separate decoders or heads that take the encoded vector and output (a) the inferred previous state of the cell, and (b) the vector representation of the CA rule. The rule's representation could be a fixed-size vector that you map to specific CA rules, either through a classification layer or some form of regression, depending on how you encode the rules.

### Training Strategy

- **Forward Training:** Use a dataset of 3-bit vectors with their corresponding next states under various CA rules. Train the forward pathway to minimize the prediction error (e.g., binary cross-entropy loss for predicting the next state as 0 or 1).
- **Backward Training:** Use sequences of cell states generated by applying CA rules. The backward pathway is trained to: (1) accurately infer the previous state (binary classification for each cell state), and (2) correctly identify the CA rule used (multi-class classification or regression, depending on your rule encoding). This may involve a combined loss function that accounts for both tasks.

### Implementation in PyTorch

Here's a simplified example of how you might start implementing such a model in PyTorch:

```python
import torch
import torch.nn as nn
import torch.optim as optim

class DualDirectionalCANetwork(nn.Module):
    def __init__(self, num_rules):
        super(DualDirectionalCANetwork, self).__init__()
        # Forward pathway
        self.forward_net = nn.Sequential(
            nn.Linear(3, 10),  # Example sizes
            nn.ReLU(),
            nn.Linear(10, 1),
            nn.Sigmoid()
        )
        # Backward pathway
        self.encoder = nn.LSTM(input_size=1, hidden_size=20, batch_first=True)
        self.decoder_state = nn.Linear(20, 1)
        self.decoder_rule = nn.Linear(20, num_rules)
    
    def forward(self, x, sequence=None):
        if sequence is None:
            # Forward prediction
            return self.forward_net(x)
        else:
            # Backward inference
            _, (hidden, _) = self.encoder(sequence)
            prev_state = self.decoder_state(hidden.squeeze(0))
            rule = self.decoder_rule(hidden.squeeze(0))
            return prev_state, rule

# Example usage
num_rules = 256  # For elementary CA, there are 256 possible rules
model = DualDirectionalCANetwork(num_rules=num_rules)

# Example forward input
forward_input = torch.tensor([[1, 0, 1]], dtype=torch.float)
forward_output = model(forward_input)

# Example backward input (sequence of states)
sequence_input = torch.rand((1, 10, 1))  # Example: batch_size=1, sequence_length=10
prev_state, rule = model(None, sequence_input)

# Define loss functions and optimizer
# You would need to customize these based on your specific requirements and data
```

This code outlines the basic structure of the model and how you might implement the forward and backward pathways. You'll need to refine the architecture, loss functions, and training procedure based on your specific requirements, the complexity of the CA rules you're working with, and the characteristics of your data.

**Comment by csmangum** [comment]
![image](https://github.com/csmangum/GCA/assets/30098247/fd09d93c-8122-4ad7-ba02-f9ac5f51a476)

The diagram above illustrates the structure of a Dual-Directional (DD) Model. It showcases two primary pathways:

- **Forward Pathway:** This part of the model takes an "Input State" and predicts the "Next State," following the forward dynamics of the system.
  
- **Backward Pathway:** Conversely, this pathway works with a "State Sequence" to infer the "Previous State & Rule," effectively understanding the system in reverse.

Each pathway is designed to handle different aspects of the problem, with the forward pathway focusing on prediction based on the current state and the backward pathway dedicated to inferring past states and the rules that led to the current situation. This dual approach allows for a comprehensive understanding and manipulation of the system's dynamics.

**Comment by csmangum** [comment]
Right now not going to try and have some type of weight sharing. Will start with separated pathways and see how that performs.
